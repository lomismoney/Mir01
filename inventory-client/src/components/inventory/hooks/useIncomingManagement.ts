import { useState, useMemo } from "react";
import { useRouter } from "next/navigation";
import { useStores, useAllInventoryTransactions } from "@/hooks";
import { useDebounce } from "@/hooks/use-debounce";
import { useToast } from "@/components/ui/use-toast";

interface IncomingFilters {
  store_id?: number;
  start_date?: string;
  end_date?: string;
  product_name?: string;
  order_number?: string;
  page?: number;
  per_page?: number;
}

/**
 * 入庫管理核心邏輯 Hook
 */
export function useIncomingManagement() {
  const { toast } = useToast();
  const router = useRouter();

  // 狀態管理
  const [purchaseDialogOpen, setPurchaseDialogOpen] = useState(false);
  const [filters, setFilters] = useState<IncomingFilters>({
    page: 1,
    per_page: 20,
  });
  const [productNameInput, setProductNameInput] = useState("");

  // 使用 debounce 優化商品名稱搜尋
  const debouncedProductName = useDebounce(productNameInput, 300);

  // 獲取門市列表
  const { data: storesData, isLoading: isLoadingStores } = useStores();

  // 獲取入庫歷史（只查詢 addition 類型的交易記錄）
  const {
    data: transactionsData,
    isLoading: isLoadingTransactions,
    error: transactionsError,
    refetch: refetchTransactions,
  } = useAllInventoryTransactions({
    type: "addition", // 只查詢入庫記錄
    store_id: filters.store_id,
    start_date: filters.start_date,
    end_date: filters.end_date,
    product_name: debouncedProductName || undefined,
    page: filters.page,
    per_page: filters.per_page,
  });

  /**
   * 處理門市篩選變更
   */
  const handleStoreChange = (value: string) => {
    const storeId = value === "all" ? undefined : parseInt(value);
    setFilters((prev) => ({
      ...prev,
      store_id: storeId,
      page: 1,
    }));
  };

  /**
   * 處理日期篩選變更
   */
  const handleDateChange = (
    field: "start_date" | "end_date",
    value: string,
  ) => {
    setFilters((prev) => ({
      ...prev,
      [field]: value || undefined,
      page: 1,
    }));
  };

  /**
   * 重置所有篩選器
   */
  const handleResetFilters = () => {
    setFilters({
      page: 1,
      per_page: 20,
    });
    setProductNameInput("");
  };

  /**
   * 刷新數據
   */
  const handleRefresh = () => {
    refetchTransactions();
    toast({
      title: "重新整理",
      description: "已重新載入入庫數據",
    });
  };

  /**
   * 計算當前篩選器的數量
   */
  const getActiveFiltersCount = () => {
    let count = 0;
    if (debouncedProductName) count++;
    if (filters.store_id) count++;
    if (filters.start_date) count++;
    if (filters.end_date) count++;
    return count;
  };

  /**
   * 分頁處理
   */
  const handlePageChange = (newPage: number) => {
    setFilters((prev) => ({
      ...prev,
      page: newPage,
    }));
  };

  return {
    // 狀態
    purchaseDialogOpen,
    setPurchaseDialogOpen,
    filters,
    productNameInput,
    setProductNameInput,
    
    // 數據
    storesData,
    transactionsData,
    isLoadingStores,
    isLoadingTransactions,
    transactionsError,
    
    // 處理函數
    handleStoreChange,
    handleDateChange,
    handleResetFilters,
    handleRefresh,
    getActiveFiltersCount,
    handlePageChange,
    refetchTransactions,
    
    // 工具
    router,
    toast,
  };
}