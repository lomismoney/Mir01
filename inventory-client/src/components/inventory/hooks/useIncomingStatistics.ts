import { useMemo } from "react";

/**
 * 入庫統計計算 Hook
 */
export function useIncomingStatistics(transactionsData: any) {
  const statistics = useMemo(() => {
    if (!transactionsData?.data) {
      return {
        todayCount: 0,
        weekCount: 0,
        totalCount: 0,
        pendingCount: 0,
      };
    }

    const transactions = transactionsData.data;
    const today = new Date().toDateString();
    const weekAgo = new Date(Date.now() - 7 * 24 * 60 * 60 * 1000);

    // 今日入庫次數
    const todayCount = transactions.filter((t: any) => {
      const transactionDate = new Date(t.created_at || "").toDateString();
      return transactionDate === today;
    }).length;

    // 本週入庫次數
    const weekCount = transactions.filter((t: any) => {
      const transactionDate = new Date(t.created_at || "");
      return transactionDate >= weekAgo;
    }).length;

    // 總入庫次數
    const totalCount = transactions.length;

    // 待處理入庫（如果有狀態字段的話）
    const pendingCount = transactions.filter((t: any) => 
      t.status === "pending" || t.status === "processing"
    ).length;

    return {
      todayCount,
      weekCount,
      totalCount,
      pendingCount,
    };
  }, [transactionsData]);

  return statistics;
}