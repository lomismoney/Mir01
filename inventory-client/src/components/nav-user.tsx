"use client";

import { useState, useEffect } from "react";
import Link from "next/link";
import {
  CreditCard,
  MoreVertical,
  LogOut,
  Bell,
  UserCircle,
} from "lucide-react";

import { Avatar, AvatarFallback } from "@/components/ui/avatar";
import {
  DropdownMenu,
  DropdownMenuContent,
  DropdownMenuGroup,
  DropdownMenuItem,
  DropdownMenuLabel,
  DropdownMenuSeparator,
  DropdownMenuTrigger,
} from "@/components/ui/dropdown-menu";
import {
  SidebarMenu,
  SidebarMenuButton,
  SidebarMenuItem,
  useSidebar,
} from "@/components/ui/sidebar";
import { Skeleton } from "@/components/ui/skeleton";
import { useSession, signOut } from "next-auth/react";
import { clearTokenCache } from "@/lib/apiClient";

/**
 * 用戶導覽元件（Auth.js + 高性能緩存整合版本）
 *
 * 功能特色：
 * - 顯示當前登入用戶的資訊和選單選項
 * - 使用 Auth.js useSession Hook 獲取用戶狀態
 * - 整合智能 token 緩存管理，登出時自動清理
 * - 確保系統性能優化的完整性
 * - 修復 Next.js Hydration 錯誤
 */
export function NavUser() {
  const { isMobile } = useSidebar();
  const { data: session, status } = useSession();

  // 🚀 修復 Hydration 錯誤：延遲獲取狀態
  const [mounted, setMounted] = useState(false);

  useEffect(() => {
    setMounted(true);
  }, []);

  // 從 Auth.js session 中提取用戶資訊和狀態
  const user = session?.user;
  const isLoading = status === "loading";

  /**
   * 處理登出點擊事件（高性能版本）
   *
   * 優化特性：
   * 1. 清理 API 客戶端的 token 緩存
   * 2. 確保下次登入時重新獲取 token
   * 3. 完整的登出流程整合
   */
  const handleLogout = () => {
    // 🧹 清理 token 緩存，確保安全登出
    clearTokenCache();

    // 🚪 執行 Auth.js 登出流程
    signOut({ callbackUrl: "/login" });
  };

  // 🎯 在客戶端 hydration 完成前，顯示骨架屏避免不一致
  if (!mounted || isLoading) {
    return (
      <SidebarMenu>
        <SidebarMenuItem>
          <SidebarMenuButton
            size="lg"
            disabled
            suppressHydrationWarning
           
          >
            <Skeleton className="h-8 w-8 rounded-lg" />
            <div
              className="grid flex-1 text-left text-sm leading-tight"
             
            >
              <Skeleton className="h-4 w-20 mb-1" />
              <Skeleton className="h-3 w-24" />
            </div>
            <Skeleton className="ml-auto h-4 w-4" />
          </SidebarMenuButton>
        </SidebarMenuItem>
      </SidebarMenu>
    );
  }

  // 如果沒有用戶資料，顯示預設狀態
  if (!user) {
    return (
      <SidebarMenu>
        <SidebarMenuItem>
          <SidebarMenuButton
            size="lg"
            disabled
            suppressHydrationWarning
           
          >
            <Avatar className="h-8 w-8 rounded-lg grayscale">
              <AvatarFallback className="rounded-lg">
                ?
              </AvatarFallback>
            </Avatar>
            <div
              className="grid flex-1 text-left text-sm leading-tight"
             
            >
              <span
                className="truncate font-medium text-muted-foreground"
               
              >
                未登入
              </span>
              <span
                className="text-muted-foreground truncate text-xs"
               
              >
                請先登入
              </span>
            </div>
          </SidebarMenuButton>
        </SidebarMenuItem>
      </SidebarMenu>
    );
  }

  return (
    <SidebarMenu>
      <SidebarMenuItem>
        <DropdownMenu>
          <DropdownMenuTrigger asChild>
            <SidebarMenuButton
              size="lg"
              className="data-[state=open]:bg-sidebar-accent data-[state=open]:text-sidebar-accent-foreground"
              suppressHydrationWarning
             
            >
              <Avatar
                className="h-8 w-8 rounded-lg grayscale"
               
              >
                <AvatarFallback className="rounded-lg">
                  {user?.name?.charAt(0).toUpperCase() || "U"}
                </AvatarFallback>
              </Avatar>
              <div
                className="grid flex-1 text-left text-sm leading-tight"
               
              >
                <span className="truncate font-medium">
                  {user?.name || "未知用戶"}
                </span>
                <span
                  className="text-muted-foreground truncate text-xs"
                 
                >
                  {user?.username || "無帳號資訊"}
                </span>
              </div>
              <MoreVertical className="ml-auto h-4 w-4" />
            </SidebarMenuButton>
          </DropdownMenuTrigger>
          <DropdownMenuContent
            className="w-(--radix-dropdown-menu-trigger-width) min-w-56 rounded-lg"
            // 🎯 只在客戶端 mounted 後使用 isMobile，避免 hydration 錯誤
            side={mounted && isMobile ? "bottom" : "right"}
            align="end"
            sideOffset={4}
            suppressHydrationWarning
           
          >
            <DropdownMenuLabel className="p-0 font-normal">
              <div
                className="flex items-center gap-2 px-1 py-1.5 text-left text-sm"
               
              >
                <Avatar className="h-8 w-8 rounded-lg">
                  <AvatarFallback className="rounded-lg">
                    {user?.name?.charAt(0).toUpperCase() || "U"}
                  </AvatarFallback>
                </Avatar>
                <div
                  className="grid flex-1 text-left text-sm leading-tight"
                 
                >
                  <span className="truncate font-medium">
                    {user?.name || "未知用戶"}
                  </span>
                  <span
                    className="text-muted-foreground truncate text-xs"
                   
                  >
                    {user?.username || "無帳號資訊"}
                  </span>
                </div>
              </div>
            </DropdownMenuLabel>
            <DropdownMenuSeparator />
            <DropdownMenuGroup>
              <DropdownMenuItem className="flex items-center gap-2" asChild>
                <Link href="/profile">
                  <UserCircle className="h-4 w-4" />
                  <span>個人資料</span>
                </Link>
              </DropdownMenuItem>
              <DropdownMenuItem className="flex items-center gap-2">
                <CreditCard className="h-4 w-4" />
                <span>帳戶設定</span>
              </DropdownMenuItem>
              <DropdownMenuItem className="flex items-center gap-2">
                <Bell className="h-4 w-4" />
                <span>通知設定</span>
              </DropdownMenuItem>
            </DropdownMenuGroup>
            <DropdownMenuSeparator />
            <DropdownMenuItem onClick={handleLogout} className="flex items-center gap-2">
              <LogOut className="h-4 w-4" />
              <span>登出</span>
            </DropdownMenuItem>
          </DropdownMenuContent>
        </DropdownMenu>
      </SidebarMenuItem>
    </SidebarMenu>
  );
}
