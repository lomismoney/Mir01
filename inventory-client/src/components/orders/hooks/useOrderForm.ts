import { useForm } from "react-hook-form";
import { zodResolver } from "@hookform/resolvers/zod";
import * as z from "zod";
import { useEffect } from "react";

// 🎯 使用 Zod 提前定義表單驗證規則
const orderFormSchema = z.object({
  customer_id: z.number().min(1, "必須選擇一個客戶"),
  shipping_address: z.string().min(1, "運送地址為必填"),
  payment_method: z.string().min(1, "必須選擇付款方式"),
  order_source: z.string().min(1, "必須選擇客戶來源"),
  shipping_status: z.string(),
  payment_status: z.string(),
  fulfillment_priority: z.enum(["normal", "high", "urgent", "low"]).optional(),
  priority_reason: z.string().optional(),
  shipping_fee: z.number().min(0).optional(),
  tax: z.number().min(0).optional(),
  discount_amount: z.number().min(0).optional(),
  notes: z.string().optional(),
  items: z
    .array(
      z.object({
        id: z.number().optional(),
        product_variant_id: z.number().nullable(),
        is_stocked_sale: z.boolean(),
        status: z.string(),
        quantity: z.number().min(1, "數量至少為 1"),
        price: z.number().min(0, "價格不能為負"),
        product_name: z.string().min(1, "商品名稱為必填"),
        sku: z.string().min(1, "SKU 為必填"),
        custom_specifications: z.record(z.string()).optional(),
        imageUrl: z.string().optional().nullable(),
        stock: z.number().optional(), // 🎯 添加庫存字段
      })
    )
    .min(1, "訂單至少需要一個品項"),
});

export type OrderFormValues = z.infer<typeof orderFormSchema>;

interface UseOrderFormProps {
  initialData?: Partial<OrderFormValues>;
  onSubmit: (values: OrderFormValues) => void;
}

export function useOrderForm({ initialData, onSubmit }: UseOrderFormProps) {
  const form = useForm<OrderFormValues>({
    resolver: zodResolver(orderFormSchema),
    defaultValues: initialData || {
      shipping_status: "pending",
      payment_status: "pending",
      fulfillment_priority: "normal",
      priority_reason: "",
      shipping_fee: 0,
      tax: 0,
      discount_amount: 0,
      items: [],
    },
  });

  // 當 initialData 變更時，重置表單
  useEffect(() => {
    if (initialData) {
      form.reset(initialData);
    }
  }, [initialData, form]);

  function handleSubmit(values: OrderFormValues) {
    const orderData: OrderFormValues = {
      ...values,
      items: values.items.map((item) => ({
        ...item,
        custom_specifications: item.custom_specifications
          ? item.custom_specifications
          : undefined,
      })),
    };

    onSubmit(orderData);
  }

  return {
    form,
    handleSubmit,
    orderFormSchema,
  };
}