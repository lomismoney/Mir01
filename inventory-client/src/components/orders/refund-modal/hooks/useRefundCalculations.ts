import { useMemo } from "react";
import { RefundFormValues } from "./useRefundForm";

interface UseRefundCalculationsProps {
  watchedItems: RefundFormValues["items"];
  orderTotal: number;
}

export function useRefundCalculations({ 
  watchedItems, 
  orderTotal 
}: UseRefundCalculationsProps) {
  
  // 簡化計算邏輯 - 使用單一 useMemo 處理所有計算
  const calculations = useMemo(() => {
    if (!watchedItems) {
      return {
        totalRefundAmount: 0,
        selectedItemsCount: 0,
        totalRefundQuantity: 0,
        refundPercentage: 0,
      };
    }

    const selectedItems = watchedItems.filter(item => item.is_selected);
    
    const totalRefundAmount = selectedItems.reduce((total, item) => {
      return total + (item.price || 0) * (item.quantity || 0);
    }, 0);

    const selectedItemsCount = selectedItems.length;
    
    const totalRefundQuantity = selectedItems.reduce((sum, item) => {
      return sum + (item.quantity || 0);
    }, 0);

    const refundPercentage = orderTotal > 0 
      ? (totalRefundAmount / orderTotal) * 100 
      : 0;

    return {
      totalRefundAmount,
      selectedItemsCount,
      totalRefundQuantity,
      refundPercentage,
    };
  }, [watchedItems, orderTotal]);

  return calculations;
}