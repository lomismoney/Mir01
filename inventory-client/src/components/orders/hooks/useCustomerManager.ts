import { useState } from "react";
import { UseFormReturn } from "react-hook-form";
import { useCreateCustomer } from "@/hooks";
import { Customer } from "@/types/api-helpers";
import { OrderFormValues } from "./useOrderForm";

interface UseCustomerManagerProps {
  form: UseFormReturn<OrderFormValues>;
}

export function useCustomerManager({ form }: UseCustomerManagerProps) {
  const [isCustomerDialogOpen, setIsCustomerDialogOpen] = useState(false);
  const createCustomerMutation = useCreateCustomer();

  const handleAddNewCustomer = () => {
    setIsCustomerDialogOpen(true);
  };

  const handleCustomerCreated = (newCustomer: Customer) => {
    if (newCustomer.id) {
      form.setValue("customer_id", newCustomer.id);
      form.setValue("shipping_address", newCustomer.contact_address || "");
    }
    setIsCustomerDialogOpen(false);
  };

  return {
    isCustomerDialogOpen,
    setIsCustomerDialogOpen,
    createCustomerMutation,
    handleAddNewCustomer,
    handleCustomerCreated,
  };
}