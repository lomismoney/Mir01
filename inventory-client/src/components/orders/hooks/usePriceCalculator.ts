import { useMemo } from "react";
import { UseFormReturn } from "react-hook-form";
import { OrderFormValues } from "./useOrderForm";

interface UsePriceCalculatorProps {
  form: UseFormReturn<OrderFormValues>;
}

export function usePriceCalculator({ form }: UsePriceCalculatorProps) {
  // 使用單一 watch 調用來獲取所有需要的值，避免多次訂閱
  const watchedValues = form.watch(["items", "shipping_fee", "tax", "discount_amount"]);
  
  const items = watchedValues[0] || [];
  const shippingFee = watchedValues[1] || 0;
  const tax = watchedValues[2] || 0;
  const discountAmount = watchedValues[3] || 0;

  // 使用 useMemo 來優化計算
  const { subtotal, grandTotal } = useMemo(() => {
    const calculatedSubtotal = items.reduce((acc, item) => {
      const itemTotal = (item.price ?? 0) * (item.quantity || 0);
      return acc + itemTotal;
    }, 0);

    const calculatedGrandTotal = Math.max(
      0,
      calculatedSubtotal + shippingFee + tax - discountAmount
    );

    return {
      subtotal: calculatedSubtotal,
      grandTotal: calculatedGrandTotal,
    };
  }, [items, shippingFee, tax, discountAmount]);

  return {
    items,
    shippingFee,
    tax,
    discountAmount,
    subtotal,
    grandTotal,
  };
}