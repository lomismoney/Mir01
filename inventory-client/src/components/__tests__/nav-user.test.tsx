import React from "react";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { NavUser } from "../nav-user";
import { SidebarProvider } from "../ui/sidebar";

/**
 * NavUser 組件測試
 * 
 * 測試範圍：
 * - 載入狀態處理
 * - 登入狀態顯示
 * - 未登入狀態顯示
 * - 下拉選單功能
 * - 登出功能
 * - Hydration 安全性
 */

// 模擬 next-auth/react
jest.mock("next-auth/react", () => ({
  useSession: jest.fn(),
  signOut: jest.fn(),
}));

// 模擬 API 客戶端
jest.mock("@/lib/apiClient", () => ({
  clearTokenCache: jest.fn(),
}));

// 模擬 use-mobile hook
jest.mock("@/hooks/use-mobile", () => ({
  useIsMobile: jest.fn(),
}));

// 取得模擬的 hooks 和函式
import { useSession, signOut } from "next-auth/react";
import { clearTokenCache } from "@/lib/apiClient";
import { useIsMobile } from "@/hooks/use-mobile";

/**
 * 測試包裝器組件，提供必要的上下文
 */
const TestWrapper = ({ children, isMobile = false }: { children: React.ReactNode; isMobile?: boolean }) => {
  // 模擬 useIsMobile hook
  (useIsMobile as jest.Mock).mockReturnValue(isMobile);
  
  return (
    <SidebarProvider>
      {children}
    </SidebarProvider>
  );
};

describe("NavUser", () => {
  let user: ReturnType<typeof userEvent.setup>;

  beforeEach(() => {
    user = userEvent.setup();
    jest.clearAllMocks();
    
    // 預設 useIsMobile 返回 false
    (useIsMobile as jest.Mock).mockReturnValue(false);
  });

  /**
   * 測試載入狀態
   */
  it("應該在載入時顯示骨架屏", () => {
    useSession.mockReturnValue({
      data: null,
      status: "loading",
    });

    const { container } = render(
      <TestWrapper>
        <NavUser />
      </TestWrapper>
    );
    
    // 檢查骨架屏元素（使用 data-slot 選擇器）
    const skeletons = container.querySelectorAll('[data-slot="skeleton"]');
    expect(skeletons.length).toBeGreaterThan(0);
  });

  /**
   * 測試未登入狀態
   */
  it("應該正確顯示未登入狀態", async () => {
    useSession.mockReturnValue({
      data: null,
      status: "unauthenticated",
    });

    render(
      <TestWrapper>
        <NavUser />
      </TestWrapper>
    );
    
    // 等待組件 mounted
    await waitFor(() => {
      expect(screen.getByText("未登入")).toBeInTheDocument();
      expect(screen.getByText("請先登入")).toBeInTheDocument();
    });
  });

  /**
   * 測試已登入狀態顯示用戶資訊
   */
  it("應該正確顯示已登入用戶資訊", async () => {
    const mockSession = {
      user: {
        name: "張三",
        username: "zhangsan",
      },
    };

    useSession.mockReturnValue({
      data: mockSession,
      status: "authenticated",
    });

    render(
      <TestWrapper>
        <NavUser />
      </TestWrapper>
    );
    
    await waitFor(() => {
      expect(screen.getByText("張三")).toBeInTheDocument();
      expect(screen.getByText("zhangsan")).toBeInTheDocument();
    });
  });

  /**
   * 測試用戶名稱首字母顯示
   */
  it("應該正確顯示用戶名稱首字母", async () => {
    const mockSession = {
      user: {
        name: "John Doe",
        username: "johndoe",
      },
    };

    useSession.mockReturnValue({
      data: mockSession,
      status: "authenticated",
    });

    render(
      <TestWrapper>
        <NavUser />
      </TestWrapper>
    );
    
    await waitFor(() => {
      // 只檢查主顯示區域的首字母（下拉選單未打開時只有一個）
      expect(screen.getByText("J")).toBeInTheDocument();
      expect(screen.getByText("John Doe")).toBeInTheDocument();
    });
  });

  /**
   * 測試無用戶名稱時的默認顯示
   */
  it("應該在沒有用戶名稱時顯示默認值", async () => {
    const mockSession = {
      user: {
        name: null,
        username: null,
      },
    };

    useSession.mockReturnValue({
      data: mockSession,
      status: "authenticated",
    });

    render(
      <TestWrapper>
        <NavUser />
      </TestWrapper>
    );
    
    await waitFor(() => {
      expect(screen.getByText("未知用戶")).toBeInTheDocument();
      expect(screen.getByText("無帳號資訊")).toBeInTheDocument();
      expect(screen.getByText("U")).toBeInTheDocument(); // 默認頭像字母
    });
  });

  /**
   * 測試下拉選單開啟
   */
  it("應該能夠開啟用戶下拉選單", async () => {
    const mockSession = {
      user: {
        name: "張三",
        username: "zhangsan",
      },
    };

    useSession.mockReturnValue({
      data: mockSession,
      status: "authenticated",
    });

    render(
      <TestWrapper>
        <NavUser />
      </TestWrapper>
    );
    
    await waitFor(() => {
      expect(screen.getByText("張三")).toBeInTheDocument();
    });

    // 點擊觸發器開啟下拉選單
    const trigger = screen.getByRole("button");
    await user.click(trigger);

    // 檢查下拉選單項目
    await waitFor(() => {
      expect(screen.getByText("個人資料")).toBeInTheDocument();
      expect(screen.getByText("帳戶設定")).toBeInTheDocument();
      expect(screen.getByText("通知設定")).toBeInTheDocument();
      expect(screen.getByText("登出")).toBeInTheDocument();
    });
  });

  /**
   * 測試登出功能
   */
  it("點擊登出應該執行登出流程", async () => {
    const mockSession = {
      user: {
        name: "張三",
        username: "zhangsan",
      },
    };

    useSession.mockReturnValue({
      data: mockSession,
      status: "authenticated",
    });

    render(
      <TestWrapper>
        <NavUser />
      </TestWrapper>
    );
    
    await waitFor(() => {
      expect(screen.getByText("張三")).toBeInTheDocument();
    });

    // 開啟下拉選單
    const trigger = screen.getByRole("button");
    await user.click(trigger);

    // 點擊登出
    await waitFor(() => {
      expect(screen.getByText("登出")).toBeInTheDocument();
    });
    
    const logoutButton = screen.getByText("登出");
    await user.click(logoutButton);

    // 檢查登出函式是否被調用
    expect(clearTokenCache).toHaveBeenCalled();
    expect(signOut).toHaveBeenCalledWith({ callbackUrl: "/login" });
  });

  /**
   * 測試移動設備下的選單位置
   */
  it("在移動設備上應該調整下拉選單位置", async () => {
    const mockSession = {
      user: {
        name: "張三",
        username: "zhangsan",
      },
    };

    useSession.mockReturnValue({
      data: mockSession,
      status: "authenticated",
    });

    render(
      <TestWrapper isMobile={true}>
        <NavUser />
      </TestWrapper>
    );
    
    await waitFor(() => {
      expect(screen.getByText("張三")).toBeInTheDocument();
    });

    // 開啟下拉選單
    const trigger = screen.getByRole("button");
    await user.click(trigger);

    // 檢查下拉選單內容存在（位置調整主要是 CSS，測試基本功能）
    await waitFor(() => {
      expect(screen.getByText("個人資料")).toBeInTheDocument();
    });
  });

  /**
   * 測試Hydration錯誤預防
   */
  it("應該使用 suppressHydrationWarning 防止 hydration 錯誤", () => {
    useSession.mockReturnValue({
      data: null,
      status: "loading",
    });

    const { container } = render(
      <TestWrapper>
        <NavUser />
      </TestWrapper>
    );
    
    // 檢查是否有相關的 button 元素
    const buttonElement = container.querySelector("button");
    expect(buttonElement).toBeInTheDocument();
  });

  /**
   * 測試載入狀態下的按鈕禁用
   */
  it("載入時按鈕應該是禁用狀態", () => {
    useSession.mockReturnValue({
      data: null,
      status: "loading",
    });

    render(
      <TestWrapper>
        <NavUser />
      </TestWrapper>
    );
    
    const button = screen.getByRole("button");
    expect(button).toBeDisabled();
  });

  /**
   * 測試未登入狀態下的按鈕禁用
   */
  it("未登入時按鈕應該是禁用狀態", async () => {
    useSession.mockReturnValue({
      data: null,
      status: "unauthenticated",
    });

    render(
      <TestWrapper>
        <NavUser />
      </TestWrapper>
    );
    
    await waitFor(() => {
      const button = screen.getByRole("button");
      expect(button).toBeDisabled();
    });
  });

  /**
   * 測試組件快照
   */
  it("應該匹配快照 - 載入狀態", () => {
    useSession.mockReturnValue({
      data: null,
      status: "loading",
    });

    const { container } = render(
      <TestWrapper>
        <NavUser />
      </TestWrapper>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  /**
   * 測試組件快照 - 已登入狀態
   */
  it("應該匹配快照 - 已登入狀態", async () => {
    const mockSession = {
      user: {
        name: "張三",
        username: "zhangsan",
      },
    };

    useSession.mockReturnValue({
      data: mockSession,
      status: "authenticated",
    });

    const { container } = render(
      <TestWrapper>
        <NavUser />
      </TestWrapper>
    );
    
    await waitFor(() => {
      expect(screen.getByText("張三")).toBeInTheDocument();
    });
    
    expect(container.firstChild).toMatchSnapshot();
  });
}); 