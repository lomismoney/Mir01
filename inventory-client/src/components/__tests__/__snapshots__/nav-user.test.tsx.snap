// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavUser 應該匹配快照 - 已登入狀態 1`] = `
<div
  class="group/sidebar-wrapper has-data-[variant=inset]:bg-sidebar flex min-h-svh w-full"
  data-slot="sidebar-wrapper"
  style="--sidebar-width: 16rem; --sidebar-width-icon: 3rem;"
>
  <ul
    class="flex w-full min-w-0 flex-col gap-1"
    data-sidebar="menu"
    data-slot="sidebar-menu"
  >
    <li
      class="group/menu-item relative"
      data-sidebar="menu-item"
      data-slot="sidebar-menu-item"
    >
      <button
        aria-expanded="false"
        aria-haspopup="menu"
        class="peer/menu-button flex w-full items-center gap-2 overflow-hidden rounded-md p-2 text-left outline-hidden ring-sidebar-ring transition-[width,height,padding] focus-visible:ring-2 active:bg-sidebar-accent active:text-sidebar-accent-foreground disabled:pointer-events-none disabled:opacity-50 group-has-data-[sidebar=menu-action]/menu-item:pr-8 aria-disabled:pointer-events-none aria-disabled:opacity-50 data-[active=true]:bg-sidebar-accent data-[active=true]:font-medium data-[active=true]:text-sidebar-accent-foreground data-[state=open]:hover:bg-sidebar-accent data-[state=open]:hover:text-sidebar-accent-foreground group-data-[collapsible=icon]:size-8! [&>span:last-child]:truncate [&>svg]:size-4 [&>svg]:shrink-0 hover:bg-sidebar-accent hover:text-sidebar-accent-foreground h-12 text-sm group-data-[collapsible=icon]:p-0! data-[state=open]:bg-sidebar-accent data-[state=open]:text-sidebar-accent-foreground"
        data-active="false"
        data-sidebar="menu-button"
        data-size="lg"
        data-slot="dropdown-menu-trigger"
        data-state="closed"
        id="radix-«ro»"
        type="button"
      >
        <span
          class="relative flex size-8 shrink-0 overflow-hidden h-8 w-8 rounded-lg grayscale"
          data-slot="avatar"
        >
          <span
            class="bg-muted flex size-full items-center justify-center rounded-lg"
            data-slot="avatar-fallback"
          >
            張
          </span>
        </span>
        <div
          class="grid flex-1 text-left text-sm leading-tight"
        >
          <span
            class="truncate font-medium"
          >
            張三
          </span>
          <span
            class="text-muted-foreground truncate text-xs"
          >
            zhangsan
          </span>
        </div>
        <svg
          class="tabler-icon tabler-icon-dots-vertical ml-auto size-4"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12 12m-1 0a1 1 0 1 0 2 0a1 1 0 1 0 -2 0"
          />
          <path
            d="M12 19m-1 0a1 1 0 1 0 2 0a1 1 0 1 0 -2 0"
          />
          <path
            d="M12 5m-1 0a1 1 0 1 0 2 0a1 1 0 1 0 -2 0"
          />
        </svg>
      </button>
    </li>
  </ul>
</div>
`;

exports[`NavUser 應該匹配快照 - 載入狀態 1`] = `
<div
  class="group/sidebar-wrapper has-data-[variant=inset]:bg-sidebar flex min-h-svh w-full"
  data-slot="sidebar-wrapper"
  style="--sidebar-width: 16rem; --sidebar-width-icon: 3rem;"
>
  <ul
    class="flex w-full min-w-0 flex-col gap-1"
    data-sidebar="menu"
    data-slot="sidebar-menu"
  >
    <li
      class="group/menu-item relative"
      data-sidebar="menu-item"
      data-slot="sidebar-menu-item"
    >
      <button
        class="peer/menu-button flex w-full items-center gap-2 overflow-hidden rounded-md p-2 text-left outline-hidden ring-sidebar-ring transition-[width,height,padding] focus-visible:ring-2 active:bg-sidebar-accent active:text-sidebar-accent-foreground disabled:pointer-events-none disabled:opacity-50 group-has-data-[sidebar=menu-action]/menu-item:pr-8 aria-disabled:pointer-events-none aria-disabled:opacity-50 data-[active=true]:bg-sidebar-accent data-[active=true]:font-medium data-[active=true]:text-sidebar-accent-foreground data-[state=open]:hover:bg-sidebar-accent data-[state=open]:hover:text-sidebar-accent-foreground group-data-[collapsible=icon]:size-8! [&>span:last-child]:truncate [&>svg]:size-4 [&>svg]:shrink-0 hover:bg-sidebar-accent hover:text-sidebar-accent-foreground h-12 text-sm group-data-[collapsible=icon]:p-0!"
        data-active="false"
        data-sidebar="menu-button"
        data-size="lg"
        data-slot="sidebar-menu-button"
        disabled=""
      >
        <div
          class="bg-accent animate-pulse h-8 w-8 rounded-lg"
          data-slot="skeleton"
        />
        <div
          class="grid flex-1 text-left text-sm leading-tight"
        >
          <div
            class="bg-accent animate-pulse rounded-md h-4 w-20 mb-1"
            data-slot="skeleton"
          />
          <div
            class="bg-accent animate-pulse rounded-md h-3 w-24"
            data-slot="skeleton"
          />
        </div>
        <div
          class="bg-accent animate-pulse rounded-md ml-auto h-4 w-4"
          data-slot="skeleton"
        />
      </button>
    </li>
  </ul>
</div>
`;
